### PR TITLE
Refactor/add to cart

### LINKED ISSUE #796: Product quantity set to 0
- go to product detail page
- click `add to basket`
- quantity on the product page will change to 0

image

### PR DESCRIPTION
**Pull request by krzysztofwolski**
I want to merge this change because...
- wanted to move more components to @next
- remove duplicated component names
- after clicking `add to basket` display cart overlay to avoid confusion if the product was really added to the cart
- improve test selectors for product details page
- add updated Plop code generator and readme instructions

### Pull Request Checklist

1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] Changes are mentioned in the changelog.

### Test Environment Config

API_URI=https://master.staging.saleor.rocks/graphql/

### CODE REVIEW

**Comment by tritao** [comment]
What is this PR missing to get merged?

Anything I can help with?

**Comment by krzysztofwolski** [comment]
I recommend trying out test env to see if all working as intended ✌️

File: package.json [hunk 1]
```diff
@@ -217,6 +217,7 @@
     "test:e2e:dev": "start-server-and-test start http://localhost:3000 cy:open",
     "extract-json-messages": "rimraf dist/locale && cross-env NODE_ENV=extract babel src 'src/**/*.{ts,tsx}' -o /dev/null",
     "transpile-messages": "node scripts/transpile-tx.js",
-    "extract-messages": "npm run extract-json-messages && npm run transpile-messages"
+    "extract-messages": "npm run extract-json-messages && npm run transpile-messages",
+    "plop": "plop"
```
Comment (dominik-zeglen) [inline package.json:221]: Running plop isn't a goal itself, generating component is ;)

```suggestion
    "generate": "plop"
```

File: plop-templates/Component/stories.tsx.hbs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import { {{pascalCase name}} } from ".";
+import { I{{pascalCase name}}Props } from "./{{pascalCase name}}";
+
+const DEFAULT_PROPS: I{{ pascalCase name }}Props = {};
+
+storiesOf("@components/{{ componentType }}/{{pascalCase name}}", module)
+.addParameters({ component: {{pascalCase name}} })
+.addDecorator(story => (
+<IntlProvider locale="en">
+  <MemoryRouter initialEntries={["/"]}>{story()}</MemoryRouter>
```
Comment (dominik-zeglen) [inline plop-templates/Component/stories.tsx.hbs:15]: Does it really belong here?
Components should get navigation callbacks from parents, if we need router in mocked environment it's a signal we're doing something wrong.

File: src/@next/components/molecules/AddToCartButton/stories.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+const DEFAULT_PROPS: IAddToCartButton = {
+  disabled: false,
+  onSubmit: () => undefined,
+};
+
+storiesOf("@components/molecules/AddToCartButton", module)
+  .addParameters({ component: AddToCartButton })
+  .addDecorator(story => (
+    <IntlProvider locale="en">
+      <MemoryRouter initialEntries={["/"]}>{story()}</MemoryRouter>
```
Comment (dominik-zeglen) [inline src/@next/components/molecules/AddToCartButton/stories.tsx:18]: It doesn't make any router calls afaik

File: src/views/Product/View.tsx [hunk 4]
```diff
@@ -44,7 +44,7 @@ const extractMeta = (product: ProductDetails_product) => ({
     },
   ],
   description: product.seoDescription || product.descriptionJson,
-  image: maybe(() => product.thumbnail.url, null),
+  image: product.thumbnail.url || null,
```
Comment (dominik-zeglen) [inline src/views/Product/View.tsx:47]: `maybe` suggests that we may not have all needed data for some time

```suggestion
  image: product?.thumbnail?.url || null,
```

**Review by dominik-zeglen — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -121,6 +121,16 @@ Cypress UI mode:
...
 ```
 
+## Creating new components
+
+All new components should follow Atomic Design Guidelines and be placed in `src/@next/components` directory.
+
+Files structure can be generated using `plop`:
+
+```
+npm run plop
```
Comment (krzysztofwolski) [inline README.md:131]:
```suggestion
npm run generate
```

File: plop-templates/Component/stories.tsx.hbs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import { MemoryRouter } from "react-router";
+import { {{pascalCase name}} } from ".";
+import { I{{pascalCase name}}Props } from "./{{pascalCase name}}";
+
+const DEFAULT_PROPS: I{{ pascalCase name }}Props = {};
+
+storiesOf("@components/{{ componentType }}/{{pascalCase name}}", module)
+.addParameters({ component: {{pascalCase name}} })
+.addDecorator(story => (
+<IntlProvider locale="en">story()</IntlProvider>
```
Comment (orzechdev) [inline plop-templates/Component/stories.tsx.hbs:14]: As `IntlProvider` is always added as decorator, I think it is better to add it just in one place in `.storybook/OutLineDecorator.jsx`, so this decoration addition would be completely removed.

**Review by dominik-zeglen — APPROVED** [review]

File: plop-templates/Component/stories.tsx.hbs (hunk 6 above)
Comment (krzysztofwolski) [reply plop-templates/Component/stories.tsx.hbs:14]: Its outside of the scope of this PR, editing existing PRs would be needed. Created Issue: https://app.clickup.com/t/2549495/SALEOR-1194

**Review by orzechdev — APPROVED** [review]

**Pull request merged**